llama-index-integrations/llms/llama-index-llms-nvidia-ai-playground/llama_index/llms/nvidia_ai_playground/base.py

```diff
@@ -0,0 +1,311 @@
+            callback_manager=callback_manager,
+            system_prompt=system_prompt,
+            messages_to_prompt=messages_to_prompt,
+            complettion_to_prompt=complettion_to_prompt,
+            pydantic_program_mode=pydantic_program_mode,
+            output_parser=output_parser,
+        )
+
+    @classmethod
+    def class_name(cls) -> str:
+        return "NvidiaAIPlayground"
+
+    @property
+    def metadata(self) -> LLMMetadata:
+        return LLMMetadata(
+            context_window=playground_modelname_to_contextsize(self.model),
+            num_output=self.max_tokens,
+            is_chat_model=True,
+            model_name=self.model,
+        )
+
+    @property
+    def _model_kwargs(self) -> Dict[str, Any]:
+        base_kwargs = {
+            "model": self.model,
+            "temperature": self.temperature,
+            "max_tokens": self.max_tokens,
+        }
+        return base_kwargs
+
+    # === Helper Methods ===
+
+    def _get_all_kwargs(self, **kwargs: Any) -> Dict[str, Any]:
+        return {
+            **self._model_kwargs,
+            **kwargs,
+        }
+
+    def _get_response_token_counts(self, raw_response: Any) -> dict:
+        """Get the token usage reported by the response."""
+        if not isinstance(raw_response, dict):
+            return {}
+
+        usage = raw_response.get("usage", {})
+        if usage is None:
+            return {}
+
+        return {
+            "prompt_tokens": usage.get("prompt_tokens", 0),
+            "completion_tokens": usage.get("completion_tokens", 0),
+            "total_tokens": usage.get("total_tokens", 0),
+        }
+
+    # === Sync Methods ===
+
+    @llm_chat_callback()
+    def chat(self, messages: Sequence[ChatMessage], **kwargs: Any) -> ChatResponse:
+        message_dicts = to_openai_message_dicts(messages)
+
+        response = self._client.chat.completions.create(
+            messages=message_dicts,
+            stream=False,
+            **self._get_all_kwargs(**kwargs),
+        )
+        playground_openai_message = response.choices[0].message
+        message = from_openai_message(playground_openai_message)
+
+        return ChatResponse(
+            message=message,
+            raw=response,
+            additional_kwargs=self._get_response_token_counts(response),
+        )
+
+    @llm_completion_callback()
+    def complete(
+        self, prompt: str, formatted: bool = False, **kwargs: Any
+    ) -> CompletionResponse:
+        complete_fn = chat_to_completion_decorator(self.chat)
+        return complete_fn(prompt, **kwargs)
+
+    @llm_chat_callback()
+    def stream_chat(
+        self, messages: Sequence[ChatMessage], **kwargs: Any
+    ) -> ChatResponseGen:
+        all_kwargs = self._get_all_kwargs(**kwargs)
+        message_dicts = to_openai_message_dicts(messages)
+
+        response = self._client.chat.completions.create(messages=message_dicts, stream=True, **all_kwargs)
+
+        def gen() -> ChatResponseGen:
```

Comment:
why do you create gen() and return gen() instead of yielding from the loop?

Reply:
This is just the pattern found in LlamaIndex - they use this `ChatResponseGen` after the fact.

https://github.com/run-llama/llama_index/blob/main/llama-index-core/llama_index/core/base/llms/types.py#L99

Reply:
This is just the pattern found in LlamaIndex - they use this `ChatResponseGen` later.

https://github.com/run-llama/llama_index/blob/main/llama-index-core/llama_index/core/base/llms/types.py#L99

(Sorry for double ping, was on my personal acct.)